pandas/core/categorical.py

```diff
@@ -316,13 +317,14 @@ def __init__(self, values, categories=None, ordered=None, dtype=None,
             if not isinstance(values, np.ndarray):
                 values = _convert_to_list_like(values)
                 from pandas.core.series import _sanitize_array
-                # On list with NaNs, int values will be converted to float. Use
-                # "object" dtype to prevent this. In the end objects will be
-                # casted to int/... in the category assignment step.
-                if len(values) == 0 or isna(values).any():
+                # By convention, empty lists result in object dtype:
+                if len(values) == 0:
                     sanitize_dtype = 'object'
                 else:
                     sanitize_dtype = None
+                null_mask = isna(values)
```

Comment:
This is adding a lot of complexity (this whole) PR, pls see if you can simplify

Reply:
Jeff, if I could simplify I would have done it already.

But notice this PR _is_ actually simplifying the code path, potentially avoiding casting data from ``list`` to ``object`` ndarray to ``int64`` ndarray (skipping the middle step). And the way missing values are treated seems to me much cleaner and clearer than before: all the type inference is just done after removing them, so there is less dtype guesswork.

Reply:
ok will have another look
can u run the category asv and report any changes

Reply:
``BENCHMARKS NOT SIGNIFICANTLY CHANGED`` - should I copypaste the entire output?

Reply:
no, just want to make sure

Reply:
*might* want to add a benchmark for categorical creation (or a long list) with 2 cases (all nulls) and some non-null.

Reply:
Done: as expected, this only makes a difference if there are many NaNs. My guess is that it's taking half the time because it's checking the "nan-ity" of each value once instead than twice.

```
      before           after         ratio
     [e1dabf37]       [b539298c]
-        66.5±1ms       30.5±0.4ms     0.46  categoricals.Categoricals.time_constructor_all_nan

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

Reply:
yep, that's fine, just wanted to be sure